From a12e8db0c42316fde7d391e4b9510fe06281561b Mon Sep 17 00:00:00 2001
From: Ilya Fedin <fedin-ilja2010@ya.ru>
Date: Thu, 10 Feb 2022 02:18:30 +0400
Subject: [PATCH] Make indicator-application hack work in flatpak

Flatpak doesn't share $XDG_RUNTIME_DIR with the host making
QSystemTrayIcon not to work on Xubuntu/Ubuntu MATE/Ubuntu Budgie.

Although, according to https://docs.flatpak.org/en/latest/sandbox-permissions.html,
it shares a subfolder.

Pick-to: 6.3 6.2 5.15
Change-Id: I2d0043fc5a4c2c51e8fa1a920f3cada3b07eba6d
---

diff --git a/src/gui/platform/unix/dbustray/qdbustrayicon.cpp b/src/gui/platform/unix/dbustray/qdbustrayicon.cpp
index 892a99d..651a2dc 100644
--- a/src/gui/platform/unix/dbustray/qdbustrayicon.cpp
+++ b/src/gui/platform/unix/dbustray/qdbustrayicon.cpp
@@ -46,6 +46,7 @@
 #include <QRect>
 #include <QLoggingCategory>
 #include <QStandardPaths>
+#include <QFileInfo>
 #include <QDir>
 #include <QMetaObject>
 #include <QMetaEnum>
@@ -76,8 +77,12 @@
 static QString iconTempPath()
 {
     QString tempPath = QStandardPaths::writableLocation(QStandardPaths::RuntimeLocation);
-    if (!tempPath.isEmpty())
+    if (!tempPath.isEmpty()) {
+        QString flatpakId = qEnvironmentVariable("FLATPAK_ID");
+        if (QFileInfo::exists(QLatin1String("/.flatpak-info")) && !flatpakId.isEmpty())
+            tempPath += QLatin1String("/app/") + flatpakId;
         return tempPath;
+    }
 
     tempPath = QStandardPaths::writableLocation(QStandardPaths::GenericCacheLocation);
 
